### ISSUE TITLE
Add support for different game modes

### ISSUE THREAD

**Issue by kneekoo** [minetest/minetest#11457]
## Problem

Issue [#2838](https://github.com/minetest/minetest_game/issues/2838) offers the full context. Short story, "creative" should not be a privilege, but a game mode, because if multiple game modes were to be added as privileges, using "/grantme all" would make it impossible for the game to decide which game mode is expected to become active.

## Proposal

The game makers should be responsible for **implementing and supporting** whatever game modes they need, while the engine should **support switching** between them. The ability to switch game modes ("gamemode") should be a privilege that can be granted by those with elevated privileges (having "privs", I suppose). Of course, players should be able to revoke this privilege for themselves.

**Comment by Wuzzy2** [comment]
Keep in mind the situation before creative became a privilege (IN MINETEST GAME, NOT MINETEST!) was even worse: Back then, in order to get creative, you had to shut down the server, activate Creative Mode and then restart it again. Because there was only a global Creative Mode.

The privilege thus was definitely an improvement. Maybe the mistake was that granting the privilege also force-activates the Creative Mode and the player can not manually turn it off. Here's an idea: The creative priv only grants the privilege to use Creative Mode but does not activate it automatically. To activate it, the player has to do something extra, like type in a command. This is similar to other privileges like `fly` which also don't force-enable flying, but only grant the *ability* to fly. Remember, we are still in Minetest Game land, not Minetest land!

Note how all of this is perfectly in the realm of games to implement. I thus reject the idea of hard-coding some strange concept of "game modes". The problem with game modes is that they are inflexible. They lump in multiple things together and this is something very specific. This is not something to have in the engine. And if you still want game modes, that's something that can be 100% implemented in games right now, e.g. by commands or whatever. The concept of game modes is too specialized for having it in an engine. Different games have different needs, not every game wants game modes, and those that do will likely have them with their own quirks. Putting it into Minetest means we will hardcode all the logic, thus reducing flexibility. I was the main developer of MineClone 2 once. As you know, Minecraft has game modes, yet I definitely never would have requested game modes from Minetest but instead do it entirely in-game.

Also, your issue doesn't even define game modes, but I am 99% certain you stole the idea from Minecraft (that in itself is not an argument to reject it tho). ;-)

Finally, this looks like 90% something that games have to figure out and not Minetest.
But there are 10% remaining, I do not want to let Minetest let entirely off the hook:

1) There is "Creative Mode: on/off" in the Esc menu. This only represents the `creative_mode` setting, not if you actually have Creative Mode as a player. So you could have “Creative Mode: off” in the Esc menu although you actually have it (e.g. by privilege). One easy solution would be to just remove this info from the Esc menu. It's pretty obvious whether Creative Mode is enbled or nor anyway. Games define "Creative Mode" subtly differently anyway, if they even use it at all.

2) There is a "Creative Mode" checkbox forced upon all games. But games should decide whether they want to allow such a setting to begin with. Here's my promising PR with already 1 approval: #11378

3) Global Creative Mode does nothing by default … Well, almost. There is that strange inconistency. Minetest does ONE thing, and it is disabling tool wear. This is utterly weird and should be removed.

But this issue touched none of these 3 points. This issue only wants game modes, and I explained why this is a bad idea, and pointed out alternatives.

My recommendation: Reject and close this isssue for above reasons.

**Comment by neoh4x0r** [comment]
@Wuzzy2  I could use the same argument (that it's a game, and not an engine, issue) to say that that privileges should be 100% handled in-game and that the engine shouldn't touch/implement them.

However, game modes can work the same way that privileges do.
1. One or more game-modes are registered by the given game
1. A client request a game-mode change, then the engine looks at the game-mode that have been registered and lets the client select one (if appropriate) and if less than two game-modes are registered, then nothing happens.

This has the benefit of allowing various games to support different modes simply by registering a game-mode.

The main reason why I think this needs to be in the engine bultins (like privileges) is because we don't need a bunch of games duplicating code (to perform the same function) nor do we need a bunch of code overriden.

**Comment by kneekoo** [comment]
@Wuzzy2, please read issue #2838 for the full context because you misunderstood my request.

> I thus reject the idea of hard-coding some strange concept of "game modes". The problem with game modes is that they are inflexible.

This is about implementing an engine-supported game mode changer that works as the privilege system. Minetest doesn't need to implement any game mode, but just have the ability to switch to a game mode that some game supports. And to be sure, some game makers will like not having to implement a game mode switcher just to turn on their creative/spectator mode.

> It's pretty obvious whether Creative Mode is enabled or not anyway.

That's wrong and you'll understand why after you read my opening post in issue #2838.

**Comment by neoh4x0r** [comment]
@kneekoo #2838, that points to a minetest issue (**Confusing documentation of mouse left/right click**).
I think you meant this one (MTG): https://github.com/minetest/minetest_game/issues/2838

**Comment by sfan5** [comment]
> However, game modes can work the same way that privileges do. 1. One or more game-modes are registered by the given game [...]

No, Wuzzy has it exactly right.
Instead of providing more features at the engine level that modders are *forced* to deal with whether they want them or not, the engine should do less.

> The main reason why I think this needs to be in the engine bultins (like privileges) is because we don't need a bunch of games duplicating code (to perform the same function) nor do we need a bunch of code overridden.

Code duplication is a valid argument albeit a weak one. You have to consider the downside of implementing one uniform system in the engine, all games will have to use it the way it is with no way to customize any parts of it. Wuzzy's explanation basically covers why this is a bad idea.

> This is about implementing an engine-supported game mode changer that works as the privilege system.

And why does this need to be in the engine? You can implement this today exactly as you'd like with zero involvement from the engine.

**Comment by kneekoo** [comment]
> Instead of providing more features at the engine level that modders are _forced_ to deal with whether they want them or not, the engine should do less.

No one is forced to use all the features of the engine. How did you come to that conclusion?

**Comment by neoh4x0r** [comment]
>  implementing one uniform system in the engine, all games will have to use it the way it is
> with no way to customize any parts of it.

That's not an accurate description of the issue.

Nothing needs to be **implemented** in the engine, the only thing it needs to provide is an API (the **specification**).
It would be up to modders to **implement** the specification to make it work for various games (eg. registering game modes, getting the current game mode, allowing the mode to be switched, etc).

For example (**the privilege system works exactly the same way** -- albeit it mixes together the implementation and specification):

1. A game (or mod) registers a set of game modes
1. A game (or mod) registers a set of privileges (these define what game modes are allowed for a player)
1. a chat command allows changing the game mode
    1. This would be a callback into the game (or mod) and allows them to handle the switching of the game mode.
    
The specification would then only be defining two builtin functions:
1. The game-mode switch function (a chat command)
1.  The game-mode onswitch callback (if not defined, does nothing)

No implementation of how it works beyond that needs to be included in the engine.

The reason I suggested putting it in the engine as a chat command (as described) above is simply for consistency across games (you use the chat command to do it); however, all implementation should be handled by the game (or mod).

**Comment by sfan5** [comment]
> No one is forced to use all the features of the engine. How did you come to that conclusion?

If you write a game that does not use gamemodes you still have potential GUI elements, chat commands or in-game indicators provided by the engine. You have to either figure out how to get them to disappear, override the parts you can or tell your users that "there's these buttons that say gamemode don't use them or pay attention because they don't work" (awful UX).
In that sense you are forced to deal with (not _use_) features that the engine provides.

> Nothing needs to be implemented in the engine,

> The specification would then only be defining two builtin functions:
> 1.   The game-mode switch function (a chat command)

You are proposing to implement a chatcommand, that is not "nothing".

**Comment by neoh4x0r** [comment]
> > No one is forced to use all the features of the engine. How did you come to that conclusion?
> 
> If you write a game that does not use gamemodes you still have potential GUI elements, chat commands or in-game indicators provided by the engine. You have to either figure out how to get them to disappear, override the parts you can or tell your users that "there's these buttons that say gamemode don't use them or pay attention because they don't work" (awful UX).
> In that sense you are forced to deal with (not _use_) features that the engine provides.
> 
> > Nothing needs to be implemented in the engine,
> 
> > The specification would then only be defining two builtin functions:
> > 
> > 1. The game-mode switch function (a chat command)
> 
> You are proposing to implement a chatcommand, that is not "nothing".

What I meant was that there wasn't a specific implementation defined -- it merely invokes a callback to the game that provides the implementation.

**Comment by rubenwardy** [comment]
I think that the cause of this is #11458, if that were fixed then the situation improves a bit. I still think that the creative priv should allow changing modes, rather than enabling creative, but that's an issue for MTG

**Comment by neoh4x0r** [comment]
@rubenwardy I think that's the way it already works in MTG.
granting creative -- enables creative mode
revoking creative -- disables creative mode

The problem becomes that fact that a normal player cannot revoke creative mode, since it needs the same privs as granting.
This means that a player that has been granted the creative priv must have another player/admin turn if off for them too.

Also as far as #11458 goes -- the esc menu's message related to "Creative" and "Damage mode" is in the context of the world-map (configured on the main menu) and is not player-specific.

**Comment by kneekoo** [comment]
> > No one is forced to use all the features of the engine. How did you come to that conclusion?
> 
> If you write a game that does not use gamemodes you still have potential GUI elements, chat commands or in-game indicators provided by the engine. You have to either figure out how to get them to disappear, override the parts you can or tell your users that "there's these buttons that say gamemode don't use them or pay attention because they don't work" (awful UX).

So you imagined an awful UX and projected that over what I proposed, instead of actually reading and understanding the issue. You know what, whatever. This is stupid. This is not talking about how game developers and users could benefit from a better engine functionality, but rather _"let's throw new proposal out the window by using poor or blatantly wrong interpretations of what was suggested"._

But I'm honestly surprised @Wuzzy2  doesn't see the value of having this, considering he understands how MineClone 2 and others games could benefit from this functionality. In his defense, though, it looks like he didn't ready the original issue, so he didn't have the context. But you were there, @sfan5, you should know what this is about on the engine side, and that the other component is on the MTG side, where I opened a specific issue. So do what you want with this, it's your baby. Cheers.

**Comment by Wuzzy2** [comment]
But I did read the original issue, and I am not convinced. Everything you suggested so far can be perfectly implemented in the game, I don't understand why this needs any engine support. I don't see any benefit that shoving it all into the engine would have over simply doing it in a mod.

The fact that Minetest incorrectly shows "Creative Mode: on" is a bug in its own right and clearly "add gamemodes" is not the only solution. My proposed solution is to just drop this message altogether because it is based on the `creative_mode` setting only and thus fundamentally incompatible with the concept of per-player Creative Mode. I also think this message isn't really needed anyway.

The reason why you had this idea in the first place is because MTG just sucks. The problem is, the `creative` privilege is not only a privilege but also enables Creative Mode at the same time. That's a design flaw and MTG's fault, that has nothing to do with Minetest. Most importantly, MTG can already perfectly fix this on its own without having to wait for engine support. Just redesign how players get Creatve Mode.
And as I already said, game modes are trivially doable in a mod, but you didn't explain why this feature MUST be a part of the engine so badly and how mods aren't enough.

Note how everything I discussed so far, everything is still 100% at the game level, and I see no loss. Everything so far is possible without an engine change. I ask you this: What is it specifically that needs the engine involved? Implementing game modes for your own games isn't even hard. Why can't you just implement game modes as a mod, where is the problem? I really don't understand.


Finally, your suggestion is actually very vague, it leaves many crucial questions open. You didn't specify what a game mode *exactly* is. You didn't specify WHY we need game modes (as opposed to other simpler solutions) in the engine. You didn't explain what is wrong with the mod solution. You didn't specify how players will, in practice, change their game mode. You didn't specify any concrete API functionality (like actual function signatures). You didn't specify how it will act with the `creative_mode` setting. You didn't specify how you would fix the "Creative Mode: on" in the Esc menu. Sorry, but that's not convincing.

**Comment by neoh4x0r** [comment]
I don't know, it seems like to me this is more about -- **oh it's not an engine issue, therefore it must be a problem for the game to deal with**.

Even though the proposed way of doing it is the same way that has been used extensively in the lua builtin functions.

If we accept the above argument (**it's not an engine issue**)  -- then we should probably get rid of a lot of lua builtins and tell the game makers that it's their problem.

Technically the privilege system is a game problem and doesn't need builtin functions for grant/revoke or register.
I could probably find more...but can't be bothered to lookup and/or list all possible examples.

> you didn't explain why this feature MUST be a part of the engine so badly and how mods aren't enough.

So....question.
Why does the engine need to implement privileges?
Why is that not a game-only problem, but this is?

----
I believe that the ability to switch to different game-modes, defined by the game, would apply to more than just MTG.
For example MTG could register a set of game modes, or a mod could do the same.

However, what happens when another mod wants to change the game mode to one of the registered values?

Granted...the following code is just an example -- it probably needs more work, **but this isn't a PR**

**In the lua builtins: /builtin/game/gamemode.lua**
```lua
core.registered_gamemodes = {}
function core.register_gamemode(mode)
  core.registered_gamemodes[mode] = mode
  if (table.getn(core.registered_gamemodes) > 1) and (core.registered_chatcommands[mode] == nil) then
   -- the chat command would need to defined (left out from brevity)
   -- only enables the chat command if there are two or more gamemodes defined and it hasn't been registered
   core.register_chatcommand("gamemode", {...})
  if (table.getn(core.registered_gamemodes) < 2) and (core.registered_chatcommands[mode] ~= nil)
   -- unregisters the chat command if there are less than two gamemodes defined and it has been registered
   core.unregisters_chatcommand("gamemode")
  end
end

core.on_switch_gamemode(mode) = nil
function core.switch_gamemode(mode)
  if mode and (table.getn(core.registered_gamemodes) > 1) then
   if core.on_switch_gamemode and core.registered_gamemodes[mode] then
    core.on_switch_gamemode(mode)
   end
  end
end
```

**In some game or mod (in init.lua):**
```lua
function init() then
 core.register_gamemode("survial")
 core.register_gamemode("creative")
 core.register_gamemode("some_mode")
 core.register_gamemode("another_mode")
 core.on_switch_gamemode = change_gamemode
end

function change_gamemode(mode)
-- game specific implementation
end

init()
```

**Then any mod (within that game) could change the game mode (through a builtin function such as:**
```lua
core.switch_gamemode("another_mode")
```

**Or the player could do so through the chatcommand** (just pseudo code)
```
/gamemode switch another_mode
```

The point of having the gamemode registration in the lua builtins is that any mod could request a specific gamemode.
Furthermore, if more than one game would like to make use of this feature -- then by putting the registration functions
in the builtins means that there isn't a need to re-invent the wheel every time.

**Comment by sfan5** [comment]
I won't repond to the rest since that would just repeat what @Wuzzy2 wrote. Particularly the question of "<i>why</i> in the engine?" needs to be answered.
One thing though:

> If we accept the above argument (it's not an engine issue) -- then we should probably get rid of a lot of lua builtins and tell the game makers that it's their problem.
>
> Technically the privilege system is a game problem and doesn't need builtin functions for grant/revoke or register.
> I could probably find more...but can't be bothered to lookup and/or list all possible examples.

I can think of a good reason: The engine uses privileges as the sole way to control client functions such as `fly`, `fast`, `noclip`, `debug`.

But even if that wasn't the case we cannot just remove privileges due to backwards compatibility guarantees.
Existence of current features is just not a good argument for adding more comparable features. There are some features -- including creative mode -- that would have never made it into the engine with the current development direction. The goal is to eventually get rid of these and not add more tech debt.

**Comment by Wuzzy2** [comment]
The reason why game modes are not the same as privileges is simple:

Privileges are something very generic and are not really tied to a specific game. Sure, games do add privileges, but Minetest also adds its own for some very basic tasks. There are many ways to use it, but it's also very simple. It makes sense to have this in the engine because telling games to roll their own privilege would end up in an interoperability nightmare. Games use privileges all the time and the privilege system has worked well for years and I also can't recall for people to have any problem with it.

Game modes, on the other hand, are something very specific and are definitely tied to a specific game. Minetest is not supposed to register any of its own game modes.
From a mod perspective, the act of switching to a specific game mode usually means you have to know the specific game already. And in practice, I think the games themselves. I don't see the need for external mods to deal with game modes frequently, if at all.

Interoperability isn't the issue here either. Game modes only make sense in the concept of one particular game. If you are a mod that wants to switch to a SPECIFIC game mode, you probably have to depend on a SPECIFIC game either. So basically engine-supported game modes are only fake flexibility. It looks flexible on surface, but in reality, you still hard-depend on the game that adds the game mode. But if you are depending on a game already, then absolutely nothing is won for the game mode API being in the engine, you could just use the game API anyway and don't lose anything. Besides, why would you want to let external mods switch the game modes anyway? Wouldn't it be better design if the game alredy ships a complete and ready-to-use system and UX already, instead of only shipping a halfplementation and asking mods to fill in the gaps? This all seems like super esoteric to justify engine or builtin code. I am also even just a "specification" because I don't like the design at all.

So, basically, game modes are super simple in nature, I give you that. You could easily do them either in the engine or a game. I don't see the benefit for developers like me to have it in the engine rather than rolling my own. There's no real functional difference between them being in the engine or in a game. The main difference is that one solution clutters the Lua API while the other one doesn't. That's the only difference. That's why I prefer the latter one.

**Comment by hecktest** [comment]
Creative is a priv in minetest_game to allow an admin or trusted user to build structures in an otherwise survival world, without having to switch the entire game into a special mode. It's useful as it is.
Server-wide modes would be more restrictive than what a game can implement, and thus worthless.

We should instead get rid of those "creative" and "damage" flags in some way if they are causing such confusion. They have no effect in games that don't respect them anyway.
- Let a game register .conf entries it wants to expose in the main menu
- Replace esc menu "game info" with a server sent string (address, game id and modlist by default, but overridable)
- Replace the server list icons with a tag system

**Comment by kneekoo** [comment]
> Everything you suggested so far can be perfectly implemented in the game, I don't understand why this needs any engine support. I don't see any benefit that shoving it all into the engine would have over simply doing it in a mod.

Supporting something that can be used by multiple games makes the engine more appealing to game makers because they have to work less, by having a reliable implementation.

> The fact that Minetest incorrectly shows "Creative Mode: on" is a bug in its own right and clearly "add gamemodes" is not the only solution.

That's a bug that has to be dealt with on its own. This is about something else.

> The reason why you had this idea in the first place is because MTG just sucks.

It makes no difference how an idea is born. MTG can use this feature, MineClone 2 can use this feature, other games can use it.

> The problem is, the `creative` privilege is not only a privilege but also enables Creative Mode at the same time. That's a design flaw and MTG's fault, that has nothing to do with Minetest.

If Minetest had what is suggested here, MTG could've used it instead for a better result than what is currently implemented, and so could other games.

> Most importantly, MTG can already perfectly fix this on its own without having to wait for engine support. Just redesign how players get Creatve Mode.
> And as I already said, game modes are trivially doable in a mod, but you didn't explain why this feature MUST be a part of the engine so badly and how mods aren't enough.

A gamemode switcher is useful for game makers. Abstracting and standardizing it on the engine side makes it easier to deal with it. Developers can also use this knowledge to contribute to any other games based on Minetest.

> I ask you this: What is it specifically that needs the engine involved? Implementing game modes for your own games isn't even hard. Why can't you just implement game modes as a mod, where is the problem? I really don't understand.

@neoh4x0r explained it well. I'm surprised how well he understood this, compared to how badly everyone else misunderstood it. His input should've been more than enough, but it looks like it isn't. I have no clue how that's even possible when everyone here knows English well enough. Other than talking to flat earthers, I swear I haven't been so puzzled about how someone can misunderstand stuff so badly while being told how they got it wrong and what is actually going on.

> Finally, your suggestion is actually very vague, it leaves many crucial questions open. You didn't specify what a game mode _exactly_ is. You didn't specify WHY we need game modes (as opposed to other simpler solutions) in the engine. You didn't explain what is wrong with the mod solution. You didn't specify how players will, in practice, change their game mode.

By paying attention to what I said in my initial post you'd see how I actually answered each point you touched there:
- (stated in clear) the game makers should be responsible for implementing and supporting whatever game modes they need (no need to define what they might have in mind)
- (stated in clear) switching game modes is similar to privileges (nothing fancy, nothing requiring overrides anywhere)
- (implied/obvious) being similar to privileges, it's abstract and simple enough to be in the engine
- (implied/obvious) there's no need for mods for something simple that can be standardized for use by anyone who needs it
- (implied/obvious) live privileges, people issue chat commands

What I didn't mention, although @neoh4x0r did a great job and there was no need to repeat it (but I liked his comment), was that `/gamemode some_mode` is something a user issues as a command, and it works if they have the privilege, or it won't work. So that's the only thing the engine would have to do for user interaction. The games themselves would have to register their modes and what happens when the users switch. It's their implementation, nothing hardcoded in the engine. That's why it's similar to the privilege system.

> You didn't specify any concrete API functionality (like actual function signatures). You didn't specify how it will act with the `creative_mode` setting.

I'm not familiar with the engine - at all -, which is why I think limiting myself to opening a discussion and offering details on request is better than trying to suggest things outside my area of knowledge. But as I said, @neoh4x0r did a good job at offering details.

> You didn't specify how you would fix the "Creative Mode: on" in the Esc menu.

Because even if I knew how, which I don't, that's a different issue.

> I don't see the need for external mods to deal with game modes frequently, if at all.
>
> Interoperability isn't the issue here either. Game modes only make sense in the concept of one particular game. If you are a mod that wants to switch to a SPECIFIC game mode, you probably have to depend on a SPECIFIC game either. So basically engine-supported game modes are only fake flexibility. It looks flexible on surface, but in reality, you still hard-depend on the game that adds the game mode. But if you are depending on a game already, then absolutely nothing is won for the game mode API being in the engine, you could just use the game API anyway and don't lose anything. Besides, why would you want to let external mods switch the game modes anyway?

So first you thought I should've stated _"what is wrong with the mod solution"_ and now you're against it. Honestly, I can't tell if you're trolling me but this is rather depressing.

I also can't see how you concluded that _"Game modes only make sense in the concept of one particular game."_ We already have two games that would benefit from it: Minetest Game and MineClone 2. And yes, the games have to deal with whatever they want from game modes. The engine needs a switcher and callbacks, plus a function to register game modes. And it might be a good idea to have a function that queries of registered modes, so games and mods can display them however they want in a UI.

> Game modes, on the other hand, are something very specific and are definitely tied to a specific game. Minetest is not supposed to register any of its own game modes.

Games modes are tied to games. But Minetest is not a game and it doesn't have to register any mode on its own. Minetest only needs the functionality for games to use in their implementations.

> So, basically, game modes are super simple in nature, I give you that. You could easily do them either in the engine or a game. I don't see the benefit for developers like me to have it in the engine rather than rolling my own. There's no real functional difference between them being in the engine or in a game.

First of all, not every developer is knowledgeable and proficient like you and other Minetest contributors are. That's the first reason to include game mode support in the engine. Make it robust enough and document it, and people won't even have the chance to make a mess in their Lua implementations. Adding to this, maybe you like implementing everything from scratch, but not every developer does. 

The second reason is like a chicken and egg issue. By not having this feature game makers have to go through the trouble of rolling their own solutions, which is yet another thing on a long checklist that they have to complete for a full game. So they tend to prioritize other features. And I think you understand very well how game makers who pick some engine want to write less code, and rely more on core functionality, so they can get results faster and worry less about their skills.

Now maybe some of you felt insulted in some way. I say you shouldn't. I'm actually baffled how nothing seems to get through you guys, when both me and @neoh4x0r explained several times that we don't expect hardcoded stuff in the engine, but something that doesn't force the developers to use or override it.

**Comment by sfan5** [comment]
> So first you thought I should've stated "what is wrong with the mod solution" and now you're against it. Honestly, I can't tell if you're trolling me but this is rather depressing.

Your attitude is *way* off. When you don't misunderstand something you wonder if you're being "trolled" but when others don't understand something you implicitly suggest their English ability is lacking and compare them to flat earthers.
This is just rude and entirely unnecessary.

> Now maybe some of you felt insulted in some way. I say you shouldn't.

Oh great then it's not a problem at all. (sarcasm)

<hr />

> I also can't see how you concluded that "Game modes only make sense in the concept of one particular game." We already have two games that would benefit from it: Minetest Game and MineClone 2.

I'll try to explain the point Wuzzy2 was making:
You have a hypothetical mod that wants to switch between survival and spectator mode, now you want to call `minetest.switch_gamemode("spectator")`. But wait, MTG would not have a survival mode so that wouldn't work there.
At some later point you'll call `minetest.switch_gamemode("survival")`. But wait, MTG calls that "default" and not "survival" (\*) so again your code won't do what you need.
If your code only ever works with MCL2 anyway, what difference would it make to call `mcl2_core.switch_mode(...)` instead? The answer is none.

\* Example for the sake of the argument, consider that such problems can exist when you support just 2 games. Now imagine more.

With that established the thought goes this way:
Mods need to depend on the specific game to use the API anyway. => Mods can directly use API defined by the game. => The engine doesn't need gamemodes.

**Comment by neoh4x0r** [comment]
> With that established the thought goes this way:
> Mods need to depend on the specific game to use the API anyway. => Mods can directly use API defined by the game. => The engine doesn't need gamemodes.

This hypothetical situation only applies to mods that are written for a specific game (ones that would not work correctly in another game) and does not apply to mods that are game-agnostic.

If we considered these types of mods -- ones that are used by more than one game -- then (hypothetically speaking) if the API implementation is left up to each game, it could be a nightmare for a modder.

A modder would have to potentially maintain code to handle specific implementations on per-game basis -- not to mention the possibility that a particular game might not expose the API to mods (eg. not able to get a list or length of registered gamemodes, etc). All of which could be solved by having the API in the established in the builtin lua files (i.e. in the engine).

If the API was in the engine, then a modder simply uses the global API to query for registered gamemodes (to simply check the length or if nil before using it) and then does whatever they intended to do without needing to work with different API implementations. 

**If you still don't think it should go in to the engine**....

Currently what I am hearing is that this is not an engine issue and that only games should implement such an API.
However, that comes with stating "I/[we] have not adequately described an engine API for this purpose" -- that is pointless if you have already made your mind up that is a game-only issue.
 
So.....please describe to me an engine-based lua api that would be acceptable.

**Comment by kneekoo** [comment]
> When you don't understand something you wonder if you're being "trolled" but when others don't understand something you implicitly suggest their English ability is lacking and compare them to flat earthers.
> This is just rude and entirely unnecessary.

The following makes it clear that I don't challenge anyone's English knowledge in this thread:
> I have no clue how that's even possible when **everyone here knows English well enough**.

The following doesn't place anyone here on the same level with flat earthers, but clearly states than I haven't experienced communicating this badly with someone else:
> Other than talking to flat earthers, I swear I haven't been so puzzled about how someone can misunderstand stuff so badly while being told how they got it wrong and what is actually going on.

The following shows once again how you guys are stuck with the wrong assumptions about this topic:
> > I also can't see how you concluded that "Game modes only make sense in the concept of one particular game." We already have two games that would benefit from it: Minetest Game and MineClone 2.
> 
> I'll try to explain the point Wuzzy2 was making:
> You have a hypothetical mod that wants to switch between survival and spectator mode, now you want to call `minetest.switch_gamemode("spectator")`. But wait, MTG would not have a spectator mode so that wouldn't work there.
> At some later point you'll call `minetest.switch_gamemode("survival")`. But wait, MTG calls that "default" and not "survival" (*) so again your code won't do what you need.
> If your code only ever works with MCL2 anyway, what difference would it make to call `mcl2_core.switch_mode(...)` instead? The answer is none.

There's no point in imagining bad examples and use them as counterarguments. The engine is not supposed to implement any mode, nor any mode meant for some specific game. This should be easy to understand, although you keep insisting with bad examples and arguments that we haven't made.

Here comes one game, talking to the engine:
```
core.register_gamemode("survival", 1)
core.register_gamemode("creative")
```
The engine registers "survival" and sets it as default, then it also registers "creative". They get stored in a table with the registered game modes.

Either the game or a mod could then switch to a different mode: `core.switch_gamemode(another_registered_mode)`

Later, here comes the game or a mod, querying the engine for registered game modes: `core.get_gamemodes()` maybe for displaying them in a UI so the player can visually see the modes and switch by clicking, or whatever. The engine only cares about the strings stored in that gamemode table and returns all values for the game/mode to do whatever they need with them.

And if the player wants to switch modes, it is again up to the game to decide if the player has the privilege to do it, and if they do, there goes `core.switch_gamemode("some_registered_mode")` to do what the game maker decided should happen next.

When the game or a mod would have to decide what to do with some action based on the current game mode, they can check the mode and act based on it: `core.get_active_gamemode()`

All of these has nothing to do with hardcoding any game mode in the engine. Suggesting otherwise, despite having people telling you that's not what the issue is about, is counterproductive and a waste of everyone's time, including your own.

**Comment by sfan5** [comment]
> All of these has nothing to do with hardcoding any game mode in the engine. Suggesting otherwise, despite having people telling you that's not what the issue is about, is counterproductive and a waste of everyone's time, including your own.

I never suggested that so I don't know why you're bringing this up.
You make a good point though, I have better things to do than argue about a feature request that is going to be rejected anyway. I will be unsubscribing, expect no futher responses.

**Comment by kneekoo** [comment]
> Instead of providing more features at the engine level that modders are forced to deal with whether they want them or not, the engine should do less.

> You have to consider the downside of implementing one uniform system in the engine, all games will have to use it the way it is with no way to customize any parts of it.

> If you write a game that does not use gamemodes you still have potential GUI elements, chat commands or in-game indicators provided by the engine. You have to either figure out how to get them to disappear, override the parts you can or tell your users that "there's these buttons that say gamemode don't use them or pay attention because they don't work" (awful UX).
In that sense you are forced to deal with (not use) features that the engine provides.

Sure, no hardcoded stuff suggested.

**Comment by Zughy** [comment]
So, after 30 minutes of my time to read all of this (don't expect other people to do the same), considering the likes of many core devs to comments against this feature request, I proceed to close it as a "Won't add".

(for what it's worth, I think that both parties are somehow right, but unfortunately the discussion went downhill. Please let's remember there's always another human being on the other side of the screen)

**Comment by kneekoo** [comment]
There's a lot of noise and the tone here was not great, but the utility of this feature should not be disregarded as a consequence of how people carried out their conversation.

This proposal is fairly simple:
- allow game mode registration, so games don't need their own implementation
- add a command for game mode switching, allowing for game makers to do whatever they want in their games
- make `/gamemode` a privileged command, so server admins can handle it as they see fit
